crates/derive/src/stages/channel_reader.rs

```diff
@@ -1 +1,112 @@
+//! This module contains the `ChannelReader` struct.
 
+use super::channel_bank::ChannelBank;
+use crate::{
+    traits::{ChainProvider, DataAvailabilityProvider},
+    types::{Batch, BlockInfo, StageError, StageResult},
+};
+use alloc::vec::Vec;
+use alloy_rlp::Decodable;
+use anyhow::anyhow;
+use core::fmt::Debug;
+use miniz_oxide::inflate::decompress_to_vec;
+
+/// [ChannelReader] is a stateful stage that does the following:
+#[derive(Debug)]
+pub struct ChannelReader<DAP, CP>
+where
+    DAP: DataAvailabilityProvider + Debug,
+    CP: ChainProvider + Debug,
+{
+    /// The previous stage of the derivation pipeline.
+    prev: ChannelBank<DAP, CP>,
+    /// The batch reader.
+    next_batch: Option<BatchReader>,
+}
+
+impl<DAP, CP> ChannelReader<DAP, CP>
+where
+    DAP: DataAvailabilityProvider + Debug,
+    CP: ChainProvider + Debug,
+{
+    /// Create a new [ChannelReader] stage.
+    pub fn new(prev: ChannelBank<DAP, CP>) -> Self {
+        Self {
+            prev,
+            next_batch: None,
+        }
+    }
+
+    /// Pulls out the next Batch from the available channel.
+    pub async fn next_batch(&mut self) -> StageResult<Batch> {
+        if let Err(e) = self.set_batch_reader().await {
+            self.next_channel();
+            return Err(e);
+        }
+        match self
+            .next_batch
+            .as_mut()
+            .unwrap()
+            .next_batch()
+            .ok_or(anyhow!("no batch"))
+        {
+            Ok(batch) => Ok(batch),
+            Err(e) => {
+                self.next_channel();
+                Err(StageError::Custom(e))
+            }
+        }
+    }
+
+    /// Creates the batch reader from available channel data.
+    async fn set_batch_reader(&mut self) -> StageResult<()> {
+        if self.next_batch.is_none() {
+            let channel = self.prev.next_data().await?.ok_or(anyhow!("no channel"))?;
```

Comment:
The reference implementation makes a clear distinction between `io.EOF` and other errors - will need to follow that here to maintain correct error bubbling https://github.com/ethereum-optimism/optimism/blob/develop/op-node/rollup/derive/channel_in_reader.go#L70-L78

Reply:
So I believe this will bubble up the correct error because the `?` operator will return the EOF error. The `ok_or` applies to the Option which wouldn't be an EOF error in the reference implementation.

Reply:
Ah I see - you're right.